test/test_theme_data_reader.rb

```diff
@@ -0,0 +1,87 @@
+# frozen_string_literal: true
+
+require "helper"
+
+class TestThemeDataReader < JekyllUnitTest
+
+  context "site without a theme" do
+    setup do
+      @site = fixture_site("theme" => nil)
+      @reader = Reader.new(@site)
+      @reader.read_data
+      assert @site.data["greetings"]
+      assert @site.data["categories"]["dairy"]
+    end
+
+    should "should read data from source" do
+      assert_equal "Hello! I’m foo. And who are you?", @site.data["greetings"]["foo"]
+      assert_equal "Dairy", @site.data["categories"]["dairy"]["name"]
+    end
+  end
+
+  context "site with a theme without _data" do
+    setup do
+      @site = fixture_site("theme" => "test-theme-skinny")
+      @reader = Reader.new(@site)
+      @reader.read_data
+      assert @site.data["greetings"]
+      assert @site.data["categories"]["dairy"]
+    end
+
+    should "should read data from source" do
+      assert_equal "Hello! I’m foo. And who are you?", @site.data["greetings"]["foo"]
+      assert_equal "Dairy", @site.data["categories"]["dairy"]["name"]
+    end
+  end
+
+  context "site with a theme with empty _data directory" do
+    setup do
+      @site = fixture_site("theme" => "test-theme-w-empty-data")
+      @reader = Reader.new(@site)
+      @reader.read_data
+      assert @site.data["greetings"]
+      assert @site.data["categories"]["dairy"]
+    end
+
+    should "should read data from source" do
+      assert_equal "Hello! I’m foo. And who are you?", @site.data["greetings"]["foo"]
+      assert_equal "Dairy", @site.data["categories"]["dairy"]["name"]
+    end
+  end
+
+  context "site with a theme with data at root of _data" do
+    setup do
+      @site = fixture_site("theme" => "test-theme")
+      @reader = Reader.new(@site)
+      @reader.read_data
+      assert @site.data["greetings"]
+      assert @site.data["categories"]["dairy"]
+      assert @site.data["cars"]
+    end
+
+    should "should sophistical merge keys" do
+      assert_equal "Hello! I’m foo. And who are you?", @site.data["greetings"]["foo"]
+      refute_equal "Hello! I’m bar. What’s up so far?", @site.data["greetings"]["foo"]
+      assert_equal "Mercedes", @site.data["cars"]["manufacturer"]
+    end
+  end
+
+  context "site with a theme with data at root of _data and in a subdirectory" do
```

Comment:
This is not clear. What is your intention?

Reply:
Dairy in source/_data is
```yml
name: Dairy
products:
- name: cheese
  price: 5.3
- name: milk
  price: 2.5
```
Dairy in **theme** is
```yml
name: Cheese Dairy
products:
- name: spread cheese
  price: 1.2
- name: cheddar cheese
  price: 4.5
```
The test looks for the union of products under the `name` of _Dairy_.

Reply:
This is still not resolved. The title is confusing and doesn't convey your intention.